File: src/data/courses/dsa/graphs/traversals-basic-properties/dfs/code/python/dfs.py
# ============================================================
# DFS
# ============================================================
class Solution:
    def dfs(self, n, adj, start):
        result, visited = [], [False] * n
        def helper(node):
            visited[node] = True
            result.append(node)
            for nb in adj[node]:
                if not visited[nb]: helper(nb)
        helper(start)
        return result

File: src/data/courses/dsa/graphs/traversals-basic-properties/dfs/code/java/Dfs.java
// ============================================================
// DFS
// ============================================================
import java.util.*;
class Solution {
    public List<Integer> dfs(int n, List<List<Integer>> adj, int start) {
        List<Integer> result = new ArrayList<>();
        boolean[] visited = new boolean[n];
        dfsHelper(start, adj, visited, result);
        return result;
    }
    void dfsHelper(int node, List<List<Integer>> adj, boolean[] visited, List<Integer> result) {
        visited[node] = true; result.add(node);
        for (int nb : adj.get(node))
            if (!visited[nb]) dfsHelper(nb, adj, visited, result);
    }
}
